### PR TITLE
Feature/sharding/cross shard checkin

### PR DESCRIPTION
**Pull request by liayoo**
- Implement openCheckin(), closeCheckin() native functions
- Replace writeDbAndTriggerFunctions() with a real transaction

#107

### CODE REVIEW

File: tx-pool/index.js [hunk 1]
```diff
@@ -341,18 +344,25 @@ class TransactionPool {
...
-    const fullPath = db.getFullPath(parsedPath, PredefinedDbPaths.VALUES_ROOT);
-    const valueCopy = ChainUtil.isDict(action.value) ?
-        JSON.parse(JSON.stringify(action.value)) : action.value;
-    valueCopy[PredefinedDbPaths.REMOTE_TX_ACTION_RESULT] = {
-      code: success ? FunctionResultCode.SUCCESS : FunctionResultCode.FAILURE
+  doAction(action, success) {
+    const triggerTx = action.transaction;
+    const valueWithResultCode = Object.assign(
+          ChainUtil.isDict(action.value) ? action.value : { value: action.value },
+          { code: success ? FunctionResultCode.SUCCESS : FunctionResultCode.FAILURE });
```
[the remark this answers is not in the document]
Comment (platfowner) [reply tx-pool/index.js:351]: If we use the txHash included in the path, I think this can be just removed?

File: tx-pool/index.js [hunk 2]
```diff
@@ -341,18 +344,25 @@ class TransactionPool {
...
+      timestamp: triggerTx.timestamp,
+      nonce: -1
     };
-    const transaction = action.transaction;
-    db.writeDbAndTriggerFunctions(
-        parsedPath, fullPath, valueCopy, transaction.timestamp, Date.now(), transaction);
+    const ownerPrivateKey = ChainUtil.getJsObject(
+      GenesisAccounts, [AccountProperties.OWNER, AccountProperties.PRIVATE_KEY]);
+    const keyBuffer = Buffer.from(ownerPrivateKey, 'hex');
+    const endpoint = `http://localhost:${PORT}/json-rpc`;
```
Comment (platfowner) [inline tx-pool/index.js:364]: ditto

File: db/functions.js [hunk 3]
```diff
@@ -214,38 +220,152 @@ class Functions {
...
-          transaction: payloadTx.transaction,
+          timestamp: Date.now(),
+          nonce: -1
         };
-        this.tp.addRemoteTransaction(txHash, action);
-      });
-    } else {
-      logger.info(`  =>> Skip sending signed transaction to the parent blockchain: ${txHash}`);
+    // Sign and send transferTx to the node itself
+    const endpoint = `http://localhost:${PORT}/json-rpc`;
```
Comment (platfowner) [inline db/functions.js:318]: We need a todo for this?

**Review by platfowner — COMMENTED** [review]
Nice job done!
Leaving some suggestions and comments.

File: tx-pool/index.js [hunk 4]
```diff
@@ -341,18 +344,25 @@ class TransactionPool {
...
+      operation: {
+        type: WriteDbOperations.SET_VALUE,
+        ref: action.ref,
+        value: valueWithResultCode
+      },
+      timestamp: triggerTx.timestamp,
+      nonce: -1
     };
-    const transaction = action.transaction;
-    db.writeDbAndTriggerFunctions(
```
Comment (platfowner) [inline tx-pool/index.js:354]: now writeDbAndTriggerFunctions() can be removed?

File: db/functions.js (hunk 3 above)
Comment (platfowner) [reply db/functions.js:318]: Just created a PR to make the node url available from functions and tx pool: https://github.com/ainblockchain/ain-blockchain/pull/112

File: tx-pool/index.js [hunk 5]
```diff
@@ -341,18 +344,25 @@ class TransactionPool {
...
-    const parsedPath = ChainUtil.parsePath(action.ref);
-    const fullPath = db.getFullPath(parsedPath, PredefinedDbPaths.VALUES_ROOT);
-    const valueCopy = ChainUtil.isDict(action.value) ?
-        JSON.parse(JSON.stringify(action.value)) : action.value;
-    valueCopy[PredefinedDbPaths.REMOTE_TX_ACTION_RESULT] = {
-      code: success ? FunctionResultCode.SUCCESS : FunctionResultCode.FAILURE
+  doAction(action, success) {
+    const triggerTx = action.transaction;
+    const valueWithResultCode = Object.assign(
+          ChainUtil.isDict(action.value) ? action.value : { value: action.value },
```
[the remark this answers is not in the document]
Comment (liayoo) [reply tx-pool/index.js:350]: Right, haha. Then actions would be changed like so? 
as-is: { ref: `/checkin/<userAddr>/<id>/parent_finalize`, value: { txHash } }
to-be: { ref: `/checkin/<userAddr>/<id>/parent_finalize/<txHash>`, value: true } & doAction() always writes at `<actionRef>/<functionResult>` without modifying the action.value. Correct me if I misunderstood your suggestion!

File: tx-pool/index.js (hunk 4 above)
Comment (liayoo) [reply tx-pool/index.js:354]: Removed!

File: db/functions.js (hunk 3 above)
Comment (liayoo) [reply db/functions.js:318]: Thanks, will change to use the url once the pr is merged!

File: tx-pool/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (platfowner) [reply tx-pool/index.js:350]: One thing here is that 'value' actually changes depending on whether the transaction was finalized or not (failed or timed out). Maybe we can just pass value generating function e.g.

(success) => {
  return {
    tx_hash: txHash,
    result: !!success,
  };
}

?

File: tx-pool/index.js (hunk 4 above)
Comment (platfowner) [reply tx-pool/index.js:354]: Thanks!

File: db/functions.js (hunk 3 above)
Comment (platfowner) [reply db/functions.js:318]: Just merged!

File: tx-pool/index.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (liayoo) [reply tx-pool/index.js:350]: Ah, I see. I was going to change the writing path depending on the tx result, but you're saying we should keep the ref as-is, use txHash and result as the value? 
Should we then remove the value field from action?

[the remark this answers is not in the document]
Comment (platfowner) [reply tx-pool/index.js:359]: @liayoo 
Ah, I was confused a little bit, sorry.
Maybe we can just keep changing the ref by adding the tx hash with simpler value generating function like
```
(success) => {
  return !!success;
}
```
?

Yes, this function replaces the value parameter.

[the remark this answers is not in the document]
Comment (liayoo) [reply tx-pool/index.js:359]: I gotcha. I think we're now on the same page. Let me update this PR accordingly!

File: tx-pool/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (liayoo) [reply tx-pool/index.js]: done

File: tx-pool/index.js (hunk 2 above)
Comment (liayoo) [reply tx-pool/index.js]: replaced with `urlInternal`

**Comment by liayoo** [comment]
Addressed all comments, PTAL @platfowner

File: tx-pool/index.js [hunk 6]
```diff
@@ -328,31 +337,45 @@ class TransactionPool {
...
-    const valueCopy = ChainUtil.isDict(action.value) ?
-        JSON.parse(JSON.stringify(action.value)) : action.value;
-    valueCopy[PredefinedDbPaths.REMOTE_TX_ACTION_RESULT] = {
-      code: success ? FunctionResultCode.SUCCESS : FunctionResultCode.FAILURE
+  doAction(action, txHash, success) {
+    const triggerTx = action.transaction;
+    const actionTx = {
+      operation: {
+        type: WriteDbOperations.SET_VALUE,
+        ref: this._getRemoteTxActionResultPath(
```
Comment (platfowner) [inline tx-pool/index.js:365]: Can we have business logic is handled by the native functions (_openCheckin() & _closeCheckin()) as much as possible? e.g. letting the final value of 'ref' and the value function is passed by _openCheckin() so action would look like: 
```
{
  ref: '.../parent_final/<txHash>',
  valueFunction: (succes) => {
    result: !!success
  },
  transaction: ...
}
```

**Review by platfowner — APPROVED** [review]
LGTM
with some comments.
Thanks!

File: tx-pool/index.js (hunk 6 above)
Comment (liayoo) [reply tx-pool/index.js]: Thanks! That looks much better. Updated as you described :)

**Comment by liayoo** [comment]
Thanks for the review. Now merging..

**Pull request merged**